app/components/graph-editor/make-draw-node.ts

```diff
@@ -0,0 +1,111 @@
+// File: make-draw-node.ts
+// Created by: CJ Dimaano
+// Date created: January 13, 2016
+//
+// THIS FILE IS INTENDED TO BE IMPORTED ONLY INTO graph-editor.component.ts
+//
+
+
+import * as CONST from "./constants";
+import { GraphEditorCanvas, point } from "./canvas";
+import { DrawableNode } from "./drawable-interfaces";
+
+
+/**
+ * makeDrawNode  
+ *   Makes a draw function for a given node.
+ */
+export function makeDrawNode(canvas: GraphEditorCanvas, g: CanvasRenderingContext2D, n: DrawableNode, dim: any, isDragging: boolean, isHovered: boolean, pt?: point): () => void {
+    // TODO:
+    // Deal with pt.
+    let shadowColor = (isDragging ? CONST.NODE_DRAG_SHADOW_COLOR : (isHovered ? CONST.SELECTION_COLOR : undefined));
+    if (n.label && n.label.trim() !== "") {
+        let lines = n.label.split("\n");
+        ///////////////////////////
+        // Labelled, With Anchor //
+        ///////////////////////////
+        if (pt && pt !== n.position) {
+            return () => {
+                switch (n.shape) {
+                    case "circle":
+                        canvas.drawCircle(n.position, dim.r, n.borderStyle, n.borderWidth, n.borderColor, n.color, shadowColor);
+                        break;
+                    case "square":
+                        canvas.drawSquare(n.position, dim.s, n.borderStyle, n.borderWidth, n.borderColor, n.color, shadowColor);
+                        break;
+                }
+                canvas.drawText(n.position, dim.th, lines, "#fff", 2, "#000");
+                canvas.drawCircle(pt, 5, "solid", 1, "#000", "#fff");
+            };
+        }
+        /////////////////////////////
+        // Labeled, Without Anchor //
+        /////////////////////////////
+        return () => {
```

Comment:
The switch statement and drawText should be one thunk to avoid duplicate code. The only difference between the two returns in this function appears to be that the earlier one has a drawCircle in it.